### DISCUSSION TITLE
How to create a custom App for Seq - In Layman's Terms

### DISCUSSION THREAD

**Discussion by gagwithgaffer** [datalust/seq-tickets#1142] [Show and tell]
Hello All,

I decided to document how I created and installed my own custom app for Seq as I really struggled in understanding the required steps as advised by the support team. But after three days of head scratching I eventually got the result I was looking for. 

Seq's official documentation is found using the following links:

[Writing Seq Apps in C#](https://docs.datalust.co/docs/writing-seq-apps) and another one worth looking at here: [Building modern Seq Apps in C#](https://blog.datalust.co/building-modern-seq-apps-in-c-outputs/)

### Project Summary
Last year Nick from the Seq Team was very helpful in preparing a custom app for me which is available on GitHub:
[seq-app-valuelist](https://github.com/datalust/seq-app-valuelist) This works nicely, tried and tested...

The above valuelist app in principal achieves the same end goal as the the custom app I've created below, but with the difference that instead of storing the unique structured logging event property values into a text file (hidden somewhere in the burrows of Seq) we dispatch these property values off to our own Web API Endpoint as soon as they written to Seq.

### What does this Custom App do and why is it needed?

Well in my case, rather than giving the end user direct access to Seq directly, I run detailed queries from .NET Core using the Seq SDK and fetch the log events on demand into my own UI. In addition I also provide the end user a feature rich log reports form so they can run extensive queries on events of their choice. In essence by storing the list of log property values gives them an unrestricted playground in events they can search for.

The reason for creating this custom app is so we can store these property values into our own database and then query them on demand i.e. when ever the user loads the logs report form, we fetch the property values through an internal API from a backend microservice.

I originally had discussed how best to access the distinct property values with the Seq support team and it was mutually concluded that doing this through running big queries on the Seq DB would not be a scalable and affective solution, therefore a different approach here is to offload these values to another source as and when they were written to Seq.

Storing the values in our own DB has the advantage in that we can take full control of them i.e. delete or purge them whenever we want to. Now this solution might also help provide some further logs retention controls in the front end i.e. if the log properties aren't listed because they were purged, then naturally you can't use them in your queries. On this pointer, you'll later see we had stored the timestamp for these property values in SQL with this in mind for a future enhancement.

The screenshot below shows the the log report form from my front end software. (basically each drop down select box lists all the unique log property values that were logged in Seq)

![image](https://user-images.githubusercontent.com/51385414/108626749-a97e6680-7449-11eb-9ffe-a8497b9e9ed7.png)

![image](https://user-images.githubusercontent.com/51385414/108627596-670b5880-744e-11eb-8126-6199199709c6.png)

So enough of the project summary, but I thought it's important that people reading the topic understand the fuller context in what I set out to do here so the rest of my notes hopefully makes better sense...

### Project Creation Guide

We first create a new Visual Studio project, create the project as a .NET class library:

![image](https://user-images.githubusercontent.com/51385414/108626294-48559380-7447-11eb-8ff8-efcfaa3300cb.png)

![image](https://user-images.githubusercontent.com/51385414/108626377-b5692900-7447-11eb-8259-93a7853d3c46.png)

Once the project has loaded, a couple of NuGet packages were needed, I installed these through the package manager UI:

![image](https://user-images.githubusercontent.com/51385414/108628003-43e1a880-7450-11eb-8fdc-469d14f61b01.png)

Next step is we create a class that stores the app logic to manage the events for when a new log event arrives in Seq. I only required one class to be created for this app so have posted the code below in case anyone wants to adopt it:

```
using Newtonsoft.Json;
using Seq.Apps;
using Seq.Apps.LogEvents;
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace DistinctPropertyValues
{
    [SeqApp("Distinct Property Values", Description = "Track distinct values of a particular event property.")]
    public class PropertyValues : SeqApp, ISubscribeTo<LogEventData>
    {
        /// <summary>
        /// Hash Set runs for the duration of the Seq service lifetime
        /// Stores each new structured log event property value. 
        /// Value is only added once during run time so the Web API
        /// will only be called each time with new the property values.
        /// If the SEQ Service is re-started then all property values
        /// will be added to the Hash Set again. In the receiving
        /// Web API end point, the developer will need to ensure
        /// that the storage solution for these values such as SQL DB
        /// does not allow duplicate values to be stored if Seq had 
        /// been re-started. On Seq startup, The intial traffic to
        /// the Web Api Endpoint will receive a higher volume of traffic but
        /// as the Hash Set is re-populated again, the POST requests will reduce.
        /// </summary>
        readonly HashSet<string> _values = new HashSet<string>();

        // Property name is the name of the structured log event property.
        // A user configurbale text field in the Seq App settings page UI.
        [SeqAppSetting(
            DisplayName = "Property name",
            HelpText = "The name of the structured log event property to track. Only non-whitespace string values are considered.")]
        public string PropertyName { get; set; }

        // Uri base address that the app wil use to POST the property value
        // to the end user's Web API endpoint.
        // A user configurbale text field in the Seq App settings page UI.
        [SeqAppSetting(
            DisplayName = "Web API Base Uri",
            HelpText = "The Web API Endpoint base address. Example = http://localhost:8000/logproperties/addnewpropertyvalue/")]
        public string WebApiUri { get; set; }

        /// <summary>
        /// User configurable text input fields below define structure log
        /// event properties that support debugging events for the Seq App  
        /// </summary>
        [SeqAppSetting(
            DisplayName = "Log Event Category",
            HelpText = "User configurable property that is used when generating logs for the Seq App i.e. debugging information for the Web API Client")]
        public string LogEventCategory { get; set; }

        [SeqAppSetting(
            DisplayName = "Log Event Type",
            HelpText = "User configurable property that is used when generating logs for the Seq App i.e. debugging information for the Web API Client")]
        public string LogEventType { get; set; }

        [SeqAppSetting(
            DisplayName = "Log Event Source",
            HelpText = "User configurable property that is used when generating logs for the Seq App i.e. debugging information for the Web API Client")]
        public string LogEventSource { get; set; }

        /// <summary>
        /// Method is called each time a log event is processed by SEQ.
        /// </summary>
        /// <param name="evt"></param>
        public void On(Event<LogEventData> evt)
        {
            if (evt == null) throw new ArgumentNullException(nameof(evt));

            if (!evt.Data.Properties.TryGetValue(PropertyName, out var value) ||
                !(value is string s) ||
                string.IsNullOrWhiteSpace(s))
            {
                return;
            }

            // See Hash Set try add method below.
            if (TryAdd(s))
            {
                // Run a fire a forget task so we reduce the 
                //risk of a bottle neck in the Web POST request.
                Task.Run(() =>
                {
                    try
                    {
                        // Local class for property attributes 
                        // that are sent to the Web API Endpoint.
                        SeqProperty prop = new SeqProperty
                        {
                            Timestamp = evt.TimestampUtc,
                            PropertyName = PropertyName,
                            PropertyValue = s
                        };

                        using (WebClient webClient = new WebClient())
                        {
                            webClient.BaseAddress = WebApiUri;
                            webClient.Headers.Add("user-agent", "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.2; .NET CLR 1.0.3705;)");
                            webClient.Headers[HttpRequestHeader.ContentType] = "application/json";
                            string data = JsonConvert.SerializeObject(prop);
                            // Optionally specify an encoding for uploading and downloading strings.
                            webClient.Encoding = Encoding.UTF8;
                            // Upload the data.
                            string reply = webClient.UploadString(webClient.BaseAddress, data);
                            // Wait for the reply, hence why we fired up a Task to allow concurrent operations.
                            if (reply != null)
                            {
                                string logData = "New value [" + s + "] detected for property [" + PropertyName + "]. Property value succesfully sent to an external Web API with response [" + reply + "]";

                                // Log the response received back from the Web API Endpoint
                                // Write the message to Seq Logger.
                                Log.Information(
                                "{@logEventCategory}" +
                                "{@logEventType}" +
                                "{@logEventSource}",
                                LogEventCategory,
                                LogEventType,
                                LogEventSource,
                                logData);
                            }
                        }
                    }
                    catch (WebException ex)
                    {
                        String responseFromServer = ex.Message.ToString() + " ";
                        if (ex.Response != null)
                        {
                            using (WebResponse response = ex.Response)
                            {
                                Stream dataRs = response.GetResponseStream();
                                using (StreamReader reader = new StreamReader(dataRs))
                                {
                                    responseFromServer += reader.ReadToEnd();
                                }
                            }
                        }

                        string logData = "Distinct Property Value Web API returend an Error with a Server Response: " + responseFromServer;

                        // Write the error message to Seq Logger.
                        Log.Error(ex,
                        "{@logEventCategory}" +
                        "{@logEventType}" +
                        "{@logEventSource}",
                        LogEventCategory,
                        LogEventType,
                        LogEventSource,
                        logData);
                    }
                });
            }

            return;
        }

        /// <summary>
        /// Boolean returns true to the calling method 
        /// above if the structured log event property
        /// was sucesfully added to the hash set.
        /// </summary>
        /// <param name="item"></param>
        /// <returns></returns>
        public bool TryAdd(string item)
        {
            if (string.IsNullOrEmpty(item))
                throw new ArgumentException("Item cannot be null or empty.", nameof(item));

            if (_values.Contains(item))
                return false;

            _values.Add(item);
            return true;
        }
    }

    /// <summary>
    /// The structure log event attributes that 
    /// sent to the customer's own Web API Endpoint
    /// </summary>
    public class SeqProperty
    {
        public DateTime Timestamp { get; set; }
        public string PropertyName { get; set; }
        public string PropertyValue { get; set; }
    }
}
```
Below is a screenshot of the project as complete:

![image](https://user-images.githubusercontent.com/51385414/108628250-bf902500-7451-11eb-9d74-7b28ace277e1.png)

Next step is to run the build command (when updating the app I used the re-build command instead)

![image](https://user-images.githubusercontent.com/51385414/108628405-9b811380-7452-11eb-9a8c-7d6d1cb783a1.png)

![image](https://user-images.githubusercontent.com/51385414/108628430-cbc8b200-7452-11eb-8fd0-bf45683aef0b.png)

Next step was to run the 'dotnet pack' command in the package manager console:

![image](https://user-images.githubusercontent.com/51385414/108628461-0a5e6c80-7453-11eb-87d9-101a9d6cf3a8.png)

Running the above dotnet pack command creates an .nupkg file which is what we'll point Seq to when we want to upload the app into Seq:

![image](https://user-images.githubusercontent.com/51385414/108628522-5d382400-7453-11eb-9051-5c4e9f3698af.png)

As I didn't have enough experience in packaging apps for uploading/importing to/from Github/Nuget etc, I simply wanted to install the custom app via a local file upload directly into Seq via a configured 'Feed'. What confused the matter for me was when studying code templates/examples from other Seq apps on GitHub, I could see some versioning control was in force using AppVeyor in a build script.  

I actually found that preparing/creating this file wasn't needed at all in my case but simply updated the version control within Visual Studio for the project instead after every update/re-build, like the screenshot below:

**Important Note:**
When installing the app for the first time in Seq, you'll likely find everything works as expected, but every time you make a an update to the app source code in Visual Studio, you'd have to update the version number otherwise Seq wont reflect the changes you made. This caused me some real confusion and had only noticed when some of the settings in the custom app had not updated as expected following an update to the source code and re-build of the project in Visual Studio.

![image](https://user-images.githubusercontent.com/51385414/108629399-9bcfdd80-7457-11eb-91cf-8914024a631c.png)

### Installing the Custom App into Seq

First step is we have to create a local feed, I defined the file path to the Debug folder of my VS Project where the .nupkg file was created from the dotnet pack command:

![image](https://user-images.githubusercontent.com/51385414/108628821-ca988480-7454-11eb-9561-c9c052da146a.png)

![image](https://user-images.githubusercontent.com/51385414/108628859-f3b91500-7454-11eb-8e31-782af0edac26.png)

Next we upload and install the custom app (yes we still need to select the same button that implies we're loading it from NuGet)

![image](https://user-images.githubusercontent.com/51385414/108628984-bacd7000-7455-11eb-8809-c1c90c1ea5e6.png)

![image](https://user-images.githubusercontent.com/51385414/108629050-04b65600-7456-11eb-8859-515aa6980021.png)

**Important Note:**
The naming convention caught me out here too as one might assume you copy the full file name i.e. DistinctPropertyValues.1.0.0 **BUT** we actually truncate the filename by chopping off the version number at the end i.e. 'DistinctPropertyValues' otherwise you'll get an error like as shown below:

![image](https://user-images.githubusercontent.com/51385414/108629079-28799c00-7456-11eb-8d31-1ca19b379590.png)

Screenshot belows the app has successfully uploaded/installed into Seq:

![image](https://user-images.githubusercontent.com/51385414/108629122-68d91a00-7456-11eb-881a-866ac6f108f5.png)

**Another potential Gotcha!**

I was originally trying to upload the app from my W10 PC to an instance of Seq that was running in a Docker for Windows Desktop container and Seq couldn't access the file path to Windows. In my case and to save further headaches, I've just installed Seq directly onto Windows for now as I'm still developing.

Once the app is uploaded to Seq we can create one or multiple instances of the app, depending on your use case. The custom app in this write up will be deployed multiple times given each instance will take care of a single structured log event property.

Setup of the first instance:

![image](https://user-images.githubusercontent.com/51385414/108629543-52cc5900-7458-11eb-8ae5-674e9ae510d4.png)

A quick walk through the some of the settings as shown above: (some of the available settings are unique to this custom app):

- Title - simply the name of this particular app instance.
- Property Name - the name of the property we want to monitor in Seq i.e. each time a new value is logged against this structured log event property, then we'll be firing it off to our own Web API Endpoint.
- Web API Base Uri - the Uri of the web endpoint we're going to send these property values to when they arrive at Seq
- The last three text input fields (Log Event Category / Log Event Type / Log Event Source) allow us to define what property name values we want to log against when we are writing logs internally to Seq on the success or failure of this particular Web API client in our custom app. If you refer back to the source code snippet of the class we created earlier then you'll see how these string variables used.

Once the app is setup, if all is well then we should see the running status (you can see little counters that tell you how many log events per minute are being processed)

![image](https://user-images.githubusercontent.com/51385414/108629901-07b34580-745a-11eb-9aaa-7ca25d63e792.png)

### So lets see the result of the custom App :)

Screenshot below shows a log event written internally to Seq when a new log property value was recorded:

![image](https://user-images.githubusercontent.com/51385414/108629978-6ed0fa00-745a-11eb-85ab-2abb484db574.png)

In the above example, this was a log event generated and written by Serilog to Seq when I fired up my front end .NET Core 5 razor Pages App. In essence we recorded a new log event property value 'WebApp_PLUGIN' under the log property category/name of 'LogEventCategory'

In addition to the above, the important party trick here is we dispatched these values off to our own API and stored them into an SQL Table, I created a separate DB Table for each log property.

![image](https://user-images.githubusercontent.com/51385414/108630140-439ada80-745b-11eb-94c5-4261fb20c415.png)

![image](https://user-images.githubusercontent.com/51385414/108630176-63ca9980-745b-11eb-9946-e6da435a89c1.png)

The Web API Endpoint method that is called from this custom app is shown below (separate back end Net Core microservice)

![image](https://user-images.githubusercontent.com/51385414/108630235-b015d980-745b-11eb-9482-33df8cba4eb1.png)

A few more instances running nicely :)

![image](https://user-images.githubusercontent.com/51385414/108630677-da689680-745d-11eb-8bfa-2af72cb39f05.png)

### Finishing Up Notes

If you decide in deploying a custom app via Nuget/Github or some other method, I'd highly recommend studying the official topics from Seq in more detail. It's just for me I didn't fully understand what was needed and in my scenario, I was able to accomplish what was needed from the above notes.

The actual custom app I created in this example has not yet been unit tested or undergone any written tests other than simply testing it in practice, but it all seems to work nicely, just through i'd throw in the small caveat there...

I'm hoping that Seq Support team wont shoot me if I missed any vital snippets of information relevant to the required process :) but the above steps achieved what I had set out to do and they are welcome of course to add any additional info to the topic that may help others further.

A fairly long article, but one that I wrote in mind of people like myself who may not have had much experience on this topic, I really hope this helps anyone else looking to develop and test their own Seq apps.

Regards

**Comment by garryxiao** [comment]
Is it possible to use .Net Core 5 follow the similar steps?

**Reply by nblumhardt** [reply]
Hi Garry,

Yes, Seq 2021.1 and later versions will load `net5.0` libraries from app packages (although the `netstandard2.0` target will enable the app to run on many more Seq versions).

Hope this helps,
Nick

**Reply by garryxiao** [reply]
Hi @nblumhardt, that's it. Thanks a lot. Seq is a amazing product. I would like to create a App to support sending mention messages to Wechat.
